File: 6_2_2/List.cpp
#include "List.h"

void add(List *list, char x) // adds elements to the array
{
	if (list->head == nullptr)
	{
		ListElement *newElement = new ListElement;
		newElement->value = x;
		list->tail = newElement;
		list->head = newElement;
	}
	else
	{
		ListElement *newElement = new ListElement;
		ListElement *element = list->tail;
		newElement->value = x;
		if (newElement->value < list->head->value)
		{
			newElement->next = list->head;
			list->head->prev = newElement;
			list->head = newElement;
		}
		else
		{
			if (newElement->value >= list->tail->value)
			{
				list->tail->next = newElement;
				newElement->prev = list->tail;
				list->tail = newElement;
			}
			else
			{
				while (newElement->value < element->value)
				{
					element = element->prev;
				}
				ListElement *temp = element->next;
				element->next = newElement;
				newElement->prev = element;
				temp->prev = newElement;
				newElement->next = temp;
			}
		}
	}
}

void show(int counter, List *list) // shows the current array
{
	if (counter == 0)
	{
		std::cout << "The array is empty.";
	}
	else
	{
		ListElement *temp = list->head;
		while (temp != list->tail)
		{
			std::cout << temp->value << " ";
			temp = temp->next;
		}
		std::cout << temp->value << " ";
	}
	std::cout << std::endl;
}

int deleteElement(char deleteElement, int counter, List *list) // deletes the pointed element
{
	if (counter == 0)
	{
		std::cout << "The array is empty." << std::endl;
		return 0;
	}
	else
	{
		if (counter == 1)
		{
			if (list->head->value == deleteElement)
			{
				delete list->head;
				list->head = nullptr;
				return 1;
			}
			else
			{
				std::cout << "There is no this value here." << std::endl;
				return 0;
			}
		}
		else
		{
			if ((deleteElement < list->head->value) || (deleteElement > list->tail->value))
			{
				std::cout << "There is no this value here." << std::endl;
				return 0;
			}
			else
			{
				ListElement *temp = list->head;
				if (deleteElement == list->head->value)
				{
					list->head = list->head->next;
					delete list->head->prev;
					return 1;
				}
				else
				{
					if (deleteElement == list->tail->value)
					{
						list->tail = list->tail->prev;
						delete list->tail->next;
						return 1;
					}
					else
					{
						while (deleteElement > temp->value)
						{
							temp = temp->next;
						}
						if (temp->value == deleteElement)
						{
							ListElement *temp2 = temp;
							ListElement *temp1 = temp->next;
							temp->prev->next = temp1;
							temp1->prev = temp2->prev;
							delete temp2;
							return 1;
						}
						else
						{
							std::cout << "There is no this value here." << std::endl;
							return 0;
						}
					}
				}
			}
		}
	}
}

void del(List *list, int counter) // deletes the array after work
{
	ListElement *temp = list->head;
	if (counter == 0)
	{
		return;
	}
	else
	{
		while (temp != list->tail)
		{
			ListElement *temp1 = temp->next;
			delete temp;
			temp = temp1;
		}
		delete temp;
	}
}

File: homework 7_1/homework 7_1/List.cs
using System;
using System.Collections;

namespace ListAndStack
{
	/// class that can add elements to the collection, delete them and show
	public class List<T> : IEnumerable
	{
		private ListElement head;
		public int counter = 0;

		/// class that describes elements of the list
		private class ListElement
		{
			public T value { get; set; }
			public ListElement next { get; set; }

			public ListElement(T value, ListElement nextElement)
			{
				this.value = value;
				next = nextElement;
			}
		}

		/// adds element to the end of the list
		public void Push(T value)
		{
			ListElement newElement = new ListElement(value, head);
			newElement.next = head;
			head = newElement;
			counter++;
		}

		/// adds element to a position in the list
		public void AddToAPosition(T value, int position)
		{
			if (position > counter)
			{
				position = counter + 1;
			}
			if (position <= 0)
			{
				Push(value);
			}
			else
			{
				var current = head;
				for (int i = 1; i < position; ++i)
				{
					current = current.next;
				}
				var newListElement = new ListElement(value, current.next);
				current.next = newListElement;
			}
			counter++;
		}

		/// deletes element from the top of the list
		public T Pop()
		{
			if (head == null)
			{
				throw new NullElementsException("List is empty.");
			}
			else
			{
				T value = head.value;
				head = head.next;
				counter--;
				return value;
			}
		}

		/// prints elements from the list
		public void Print()
		{
			ListElement current = head;
			while (current != null)
			{
				Console.WriteLine(current.value);
				current = current.next;
			}
		}

		/// deletes element from the list
		public void DeleteElement(T value)
		{
			if (IfInTheList(value) == false)
			{
				Console.WriteLine("There is no element '{0}' in the list.", value);
				return;
			}
			ListElement current = head;
			if (current.value.Equals(value))
			{
				head = current.next;
				counter--;
				return;
			}
			while (current.next != null)
			{
				if (current.next.value.Equals(value))
				{
					current.next = current.next.next;
					counter--;
					return;
				}
				current = current.next;
			}
		}

		/// removes an element from position in the list
		public void DeleteFromThePosition(int position)
		{
			if (position >= counter)
			{
				position = counter - 1;
			}
			if (position == 0)
			{
				Pop();
				counter--;
				return;
			}
			var current = head;
			for (int i = 1; i < position; ++i)
			{
				current = current.next;
			}
			current.next = current.next.next;
			counter--;
			return;
		}

		/// checks if element is in the list
		public bool IfInTheList(T value)
		{
			ListElement current = head;
			while (current != null)
			{
				if (current.value.Equals(value))
				{
					break;
				}
				current = current.next;
			}
			return current != null;
		}

		IEnumerator IEnumerable.GetEnumerator()
		{
			return GetEnumerator();
		}

		public ListEnumerator GetEnumerator()
		{
			return new ListEnumerator(this);
		}

		public class ListEnumerator : IEnumerator
		{
			private List<T> list;
			private int position = -1;
			private ListElement current;

			public ListEnumerator(List<T> newList)
			{
				list = newList;
				current = null;
			}

			/// moves to the next list element
			public bool MoveNext()
			{
				position++;
				current = current == null ? list.head : current.next;
				return position < list.counter;
			}

			/// returns the enumerator to the position before the first element
			public void Reset()
			{
				position = -1;
				current = null;
			}

			object IEnumerator.Current
			{
				get
				{
					return Current;
				}
			}

			/// returns a value of the current element of list
			public T Current
			{
				get
				{
					try
					{
						return current.value;
					}
					catch (IndexOutOfRangeException)
					{
						throw new InvalidOperationException();
					}
				}
			}
		}
	}
}